src/vs/workbench/contrib/terminal/browser/xterm/commandNavigationAddon.ts

```diff
@@ -64,7 +64,7 @@ export class CommandNavigationAddon extends Disposable implements ICommandTracke
 		if (this._commandDetection.type === TerminalCapability.PartialCommandDetection) {
 			commands = this._commandDetection.commands;
 		} else {
-			commands = coalesce(this._commandDetection.commands.map(e => e.marker));
+			commands = coalesce(this._commandDetection.commands.filter(e => e.command).map(e => e.marker));
```

Comment:
This isn't quite what I had in mind, we want to treat empty commands specially, not ignore them completely. Consider the following example:

![image](https://user-images.githubusercontent.com/2193314/179752068-29863cde-1207-4f2e-93d1-90f29062d103.png)

---

ctrl+up should result in:

![image](https://user-images.githubusercontent.com/2193314/179752206-bfcd38af-5470-45a2-a057-601dbeae49ae.png)

This should work correctly in your change.

---

ctrl+shift+up (from the start state) should result in:

![image](https://user-images.githubusercontent.com/2193314/179752362-4b71027b-867d-490a-95cb-94a745201bbd.png)

This wouldn't work correctly yet. A more descriptive explanation of what we want here is if going from no command selected to 1 command selected, all empty commands should be skipped such that the selection starts at the boundary of the next valid command and its adjacent empty command.

We would also want this behavior to apply if moving around the buffer using ctrl+up/down and then for the first ctrl+shift+up/down. When selecting more than 1 command output it's expected to include the empty commands though.

Reply:
> When selecting more than 1 command output it's expected to include the empty commands though.

Do you mean that if the user presses another time ctr+shift+up and there is one empty command above it selects only that (image 1) or it selects every commands until the first non-empty (image 2)

Image 1:
![option2](https://user-images.githubusercontent.com/105170638/180677119-4ff320fc-b55b-49a5-aed9-70233e977aff.png)

Image 2
![option1](https://user-images.githubusercontent.com/105170638/180677129-ccd86d81-7032-4032-9e00-53f34244405a.png)

Reply:
@MonadChains I think image 2 is the best behavior here

Reply:
Any chance we can support non-contiguous row selections?

Reply:
@pfitzseb I hadn't considered that before, it would need to get implemented in xterm.js first which would need changes to how select works and gets rendered (in each of the 3 renderers)